File: proposed-projects/round-trip-experiment/reverse/20260909T181751Z_00b43b48/20260910T095901Z_ef14bd50/ported_implementation/gbnf/grammar_graph/get_serialized_rule_key.py
import json
from typing import TYPE_CHECKING

from .type_guards import is_rule_char, is_rule_char_exclude, is_rule_end, is_rule_ref

if TYPE_CHECKING:
    from .grammar_graph_types import UnresolvedRule


def get_serialized_rule_key(rule: "UnresolvedRule") -> str:
    if is_rule_end(rule):
        return "0"

    if is_rule_char(rule):
        return f"1-{json.dumps(rule.value)}"

    if is_rule_char_exclude(rule):
        return f"2-{json.dumps(rule.value)}"

    if is_rule_ref(rule):
        return f"3-{rule.value}"

    raise Exception(f"Unknown rule type: {rule!r}")

File: packages/gbnf-experiment/data/20260909T005215Z_9a3e20fe/ported_implementation/src/grammarGraph/getSerializedRuleKey.ts
import { jsonDumps, repr } from '../utils/repr.ts';
import { RuleChar, RuleCharExclude, RuleEnd } from './grammarGraphTypes.ts';
import type { UnresolvedRule } from './grammarGraphTypes.ts';
import { isRuleChar, isRuleCharExclude, isRuleEnd, isRuleRef } from './typeGuards.ts';

export const KEY_TRANSLATION = new Map<unknown, number>([
  [RuleEnd, 0],
  [RuleChar, 1],
  [RuleCharExclude, 2],
]);

export const getSerializedRuleKey = (rule: UnresolvedRule): string => {
  if (isRuleEnd(rule)) {
    return `${KEY_TRANSLATION.get(RuleEnd)}`;
  }

  if (isRuleChar(rule)) {
    return `${KEY_TRANSLATION.get(RuleChar)}-${jsonDumps(rule.value)}`;
  }

  if (isRuleCharExclude(rule)) {
    return `${KEY_TRANSLATION.get(RuleCharExclude)}-${jsonDumps(rule.value)}`;
  }

  if (isRuleRef(rule)) {
    return `3-${rule.value}`;
  }

  throw new Error(`Unknown rule type: ${repr(rule)}`);
};
